direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogObservers.java

```diff
@@ -19,35 +19,54 @@
 import cz.o2.proxima.direct.batch.BatchLogObserver.OnNextContext;
 import cz.o2.proxima.storage.Partition;
 import cz.o2.proxima.time.WatermarkSupplier;
+import cz.o2.proxima.time.Watermarks;
 
 /** Utility class related to {@link BatchLogObserver BatchLogObservers}. */
 @Internal
 public class BatchLogObservers {
 
   /**
-   * Create {@link OnNextContext} which holds watermark back on {@link Long#MIN_VALUE} until the end
-   * of data. This is the default behavior of batch readers when there is no way to time-order data.
+   * Create {@link OnNextContext} which holds watermark back on {@link Watermarks#MIN_WATERMARK}
+   * until the end of data. This is the default behavior of batch readers when there is no way to
+   * time-order data.
    *
    * @param partition the partition to create context for
    * @return a wrapped {@link OnNextContext} for given partition
    */
   public static OnNextContext defaultContext(Partition partition) {
-    return withWatermark(partition, Long.MIN_VALUE);
+    return new OnNextContext() {
+
+      @Override
+      public Partition getPartition() {
+        return partition;
+      }
+
+      @Override
+      public BatchLogObserver.Offset getOffset() {
+        throw new UnsupportedOperationException(
+            "Unable to calculate offset, because underlying data may have changed.");
```

Comment:
I'd rephrase this "because the underlying data store is not known to be immutable". That should give better hint what to do, when you encounter this exception.

Reply:
👍